### PR TITLE
Core/Spells: Fix shaman talent Ascendance (restoration)

### PR DESCRIPTION
**Pull request by aquadeus**
**Changes proposed:**

-  Fix shaman talent Ascendance (restoration)

**Tests performed:**

Tested in-game

### CODE REVIEW

File: src/server/scripts/Spells/spell_shaman.cpp [hunk 1]
```diff
@@ -213,6 +214,48 @@ class spell_sha_ancestral_guidance_heal : public SpellScript
...
+    {
+        return ValidateSpellInfo({ SPELL_SHAMAN_RESTORATIVE_MISTS });
+    }
+
+    bool CheckProc(ProcEventInfo& procInfo)
+    {
+        if (!procInfo.GetHealInfo() || !procInfo.GetSpellInfo())
+            return false;
+
+        if (procInfo.GetSpellInfo()->Id == SPELL_SHAMAN_RESTORATIVE_MISTS)
```
Comment (Nyr97) [inline src/server/scripts/Spells/spell_shaman.cpp:232]: This should be moved onto spell_proc DB table using SpellFamilyTree masks.

File: src/server/scripts/Spells/spell_shaman.cpp [hunk 2]
```diff
@@ -213,6 +214,48 @@ class spell_sha_ancestral_guidance_heal : public SpellScript
...
+    PrepareAuraScript(spell_sha_ascendance_restoration);
+
+    bool Validate(SpellInfo const* /*spellInfo*/) override
+    {
+        return ValidateSpellInfo({ SPELL_SHAMAN_RESTORATIVE_MISTS });
+    }
+
+    bool CheckProc(ProcEventInfo& procInfo)
+    {
+        if (!procInfo.GetHealInfo() || !procInfo.GetSpellInfo())
```
Comment (Nyr97) [inline src/server/scripts/Spells/spell_shaman.cpp:229]: This should be moved onto spell_proc DB table using Proc Spell Type Mask.

File: src/server/scripts/Spells/spell_shaman.cpp [hunk 3]
```diff
@@ -213,6 +214,48 @@ class spell_sha_ancestral_guidance_heal : public SpellScript
...
+            return false;
+
+        if (procInfo.GetSpellInfo()->Id == SPELL_SHAMAN_RESTORATIVE_MISTS)
+            return false;
+
+        if (!procInfo.GetHealInfo()->GetHeal())
```
Comment (Nyr97) [inline src/server/scripts/Spells/spell_shaman.cpp:235]: This should be moved onto spell_proc DB table using Proc Spell Type Mask.

File: src/server/scripts/Spells/spell_shaman.cpp [hunk 4]
```diff
@@ -213,6 +214,48 @@ class spell_sha_ancestral_guidance_heal : public SpellScript
...
+            return false;
+
+        return true;
+    }
+
+    void OnProc(AuraEffect* aurEff, ProcEventInfo& procInfo)
+    {
+        int32 heal = procInfo.GetHealInfo()->GetHeal();
+        if (heal > 0)
```
Comment (Nyr97) [inline src/server/scripts/Spells/spell_shaman.cpp:244]: Ask @Shauren if it's best to use heal > 0 get if GetHeal is EffectiveHeal (not sure if this accounts for anything that might get absorbed, we now handle heal absorbs).

Comment (mdX7) [reply src/server/scripts/Spells/spell_shaman.cpp:244]: GetHeal() is not containing the absorbed heal, e.g. by spell 365153

ref
```cpp
// this is called in void Unit::CalcHealAbsorb
void HealInfo::AbsorbHeal(uint32 amount)
{
    amount = std::min(amount, GetHeal());
    _absorb += amount;
    _heal -= amount;
    amount = std::min(amount, GetEffectiveHeal());
    _effectiveHeal -= amount;
    _hitMask |= PROC_HIT_ABSORB;
}
```

Comment (Nyr97) [reply src/server/scripts/Spells/spell_shaman.cpp:244]: Yeah, we need tests and numbers to see.

Comment (aquadeus) [reply src/server/scripts/Spells/spell_shaman.cpp:235]: on retail is healing the absorb amount, with last commit works like retail (tested both with same spell)

**Review by mdX7 — APPROVED** [review]

**Pull request merged**